File: frontend/src/lib/canvas/pipeline-boundary.ts
import type {
  ModelCatalogEntry,
  PipelineEdgeRecord,
  PipelineNodeRecord,
} from "@/lib/canvas/types";
import {
  BLOCKED_PIPELINE_MODELS,
  areWireKindsCompatible,
  getModelWireKinds,
  getModelWireLabels,
  type WireKind,
} from "@/lib/canvas/wire-types";

export type PipelineBoundaryResult = {
  valid: boolean;
  errors: string[];
  entryNodeIds: string[];
  exitNodeIds: string[];
  inputWireKind: WireKind;
  outputWireKind: WireKind;
  inputLabel: string;
  outputLabel: string;
};

function hasCycle(
  nodeIds: Set<string>,
  edges: PipelineEdgeRecord[],
): boolean {
  const inDegree = new Map<string, number>();
  const adjacency = new Map<string, string[]>();

  for (const id of nodeIds) {
    inDegree.set(id, 0);
    adjacency.set(id, []);
  }

  for (const edge of edges) {
    if (!nodeIds.has(edge.source) || !nodeIds.has(edge.target)) continue;
    adjacency.get(edge.source)?.push(edge.target);
    inDegree.set(edge.target, (inDegree.get(edge.target) ?? 0) + 1);
  }

  const queue = [...nodeIds].filter((id) => (inDegree.get(id) ?? 0) === 0);
  let visited = 0;

  while (queue.length > 0) {
    const current = queue.shift();
    if (!current) break;
    visited += 1;
    for (const next of adjacency.get(current) ?? []) {
      const degree = (inDegree.get(next) ?? 0) - 1;
      inDegree.set(next, degree);
      if (degree === 0) queue.push(next);
    }
  }

  return visited !== nodeIds.size;
}

function connectedComponents(
  nodeIds: Set<string>,
  edges: PipelineEdgeRecord[],
): number {
  const adjacency = new Map<string, Set<string>>();
  for (const id of nodeIds) adjacency.set(id, new Set());
  for (const edge of edges) {
    if (!nodeIds.has(edge.source) || !nodeIds.has(edge.target)) continue;
    adjacency.get(edge.source)?.add(edge.target);
    adjacency.get(edge.target)?.add(edge.source);
  }

  const visited = new Set<string>();
  let count = 0;

  for (const start of nodeIds) {
    if (visited.has(start)) continue;
    count += 1;
    const stack = [start];
    while (stack.length > 0) {
      const current = stack.pop();
      if (!current || visited.has(current)) continue;
      visited.add(current);
      for (const neighbor of adjacency.get(current) ?? []) {
        if (!visited.has(neighbor)) stack.push(neighbor);
      }
    }
  }

  return count;
}

function validateEdgeCompatibility(
  nodes: PipelineNodeRecord[],
  edges: PipelineEdgeRecord[],
  nodeIds: Set<string>,
): boolean {
  const modelById = new Map(nodes.map((node) => [node.id, node.modelId]));

  for (const edge of edges) {
    if (!nodeIds.has(edge.source) || !nodeIds.has(edge.target)) continue;
    const sourceModel = modelById.get(edge.source) ?? "";
    const targetModel = modelById.get(edge.target) ?? "";
    const sourceWires = getModelWireKinds(sourceModel, "", "");
    const targetWires = getModelWireKinds(targetModel, "", "");
    if (!areWireKindsCompatible(sourceWires.output, targetWires.input)) {
      return false;
    }
  }

  return true;
}

export function derivePipelineBoundaryIO(
  nodes: PipelineNodeRecord[],
  edges: PipelineEdgeRecord[],
  _modelMap?: Map<string, ModelCatalogEntry>,
): PipelineBoundaryResult {
  const errors: string[] = [];
  const empty: PipelineBoundaryResult = {
    valid: false,
    errors: ["no_nodes"],
    entryNodeIds: [],
    exitNodeIds: [],
    inputWireKind: "none",
    outputWireKind: "none",
    inputLabel: "",
    outputLabel: "",
  };

  if (nodes.length === 0) return empty;

  if (nodes.length < 2) errors.push("insufficient_nodes");

  const nodeIds = new Set(nodes.map((n) => n.id));

  for (const node of nodes) {
    if (BLOCKED_PIPELINE_MODELS.has(node.modelId)) {
      if (node.modelId === "loader/pdf" || node.modelId === "loader/image") {
        errors.push("contains_file_loader");
      } else {
        errors.push("contains_page_branch");
      }
    }
  }

  if (hasCycle(nodeIds, edges)) errors.push("cycle_detected");
  if (connectedComponents(nodeIds, edges) > 1) errors.push("disconnected_graph");
  if (!validateEdgeCompatibility(nodes, edges, nodeIds)) {
    errors.push("incompatible_connection");
  }

  const inDegree = new Map<string, number>();
  const outDegree = new Map<string, number>();
  for (const id of nodeIds) {
    inDegree.set(id, 0);
    outDegree.set(id, 0);
  }
  for (const edge of edges) {
    if (!nodeIds.has(edge.source) || !nodeIds.has(edge.target)) continue;
    outDegree.set(edge.source, (outDegree.get(edge.source) ?? 0) + 1);
    inDegree.set(edge.target, (inDegree.get(edge.target) ?? 0) + 1);
  }

  const entryNodeIds = [...nodeIds].filter((id) => (inDegree.get(id) ?? 0) === 0);
  const exitNodeIds = [...nodeIds].filter((id) => (outDegree.get(id) ?? 0) === 0);

  if (entryNodeIds.length === 0) errors.push("no_entry_node");
  if (exitNodeIds.length === 0) errors.push("no_exit_node");

  const entryInputKinds = new Set<WireKind>();
  const exitOutputKinds = new Set<WireKind>();

  for (const entryId of entryNodeIds) {
    const entryNode = nodes.find((node) => node.id === entryId);
    if (!entryNode) continue;
    const entryWires = getModelWireKinds(entryNode.modelId, "", "");
    if (entryWires.input === "none" || entryWires.input === "file") {
      errors.push("invalid_entry_input");
    } else {
      entryInputKinds.add(entryWires.input);
    }
  }

  for (const exitId of exitNodeIds) {
    const exitNode = nodes.find((node) => node.id === exitId);
    if (!exitNode) continue;
    const exitWires = getModelWireKinds(exitNode.modelId, "", "");
    if (exitWires.output === "none") {
      errors.push("invalid_exit_output");
    } else {
      exitOutputKinds.add(exitWires.output);
    }
  }

  if (entryInputKinds.size > 1) errors.push("incompatible_entry_inputs");
  if (exitOutputKinds.size > 1) errors.push("incompatible_exit_outputs");

  if (errors.length > 0) {
    const primaryEntry = nodes.find((node) => node.id === entryNodeIds[0]);
    const primaryExit = nodes.find((node) => node.id === exitNodeIds[0]);
    const entryWires = primaryEntry
      ? getModelWireKinds(primaryEntry.modelId, "", "")
      : { input: "none" as WireKind, output: "none" as WireKind };
    const exitWires = primaryExit
      ? getModelWireKinds(primaryExit.modelId, "", "")
      : { input: "none" as WireKind, output: "none" as WireKind };
    const entryLabels = primaryEntry
      ? getModelWireLabels(
          primaryEntry.modelId,
          "",
          "",
          _modelMap?.get(primaryEntry.modelId)?.category,
        )
      : { input: "", output: "" };
    const exitLabels = primaryExit
      ? getModelWireLabels(
          primaryExit.modelId,
          "",
          "",
          _modelMap?.get(primaryExit.modelId)?.category,
        )
      : { input: "", output: "" };

    return {
      valid: false,
      errors,
      entryNodeIds,
      exitNodeIds,
      inputWireKind: entryWires.input,
      outputWireKind: exitWires.output,
      inputLabel: entryLabels.input,
      outputLabel: exitLabels.output,
    };
  }

  const entryNode = nodes.find((n) => n.id === entryNodeIds[0]);
  const exitNode = nodes.find((n) => n.id === exitNodeIds[0]);
  if (!entryNode || !exitNode) {
    return { ...empty, errors: ["invalid_graph"] };
  }

  const entryCategory = _modelMap?.get(entryNode.modelId)?.category;
  const exitCategory = _modelMap?.get(exitNode.modelId)?.category;
  const entryWires = getModelWireKinds(entryNode.modelId, "", "");
  const exitWires = getModelWireKinds(exitNode.modelId, "", "");
  const entryLabels = getModelWireLabels(
    entryNode.modelId,
    "",
    "",
    entryCategory,
  );
  const exitLabels = getModelWireLabels(
    exitNode.modelId,
    "",
    "",
    exitCategory,
  );

  return {
    valid: true,
    errors: [],
    entryNodeIds,
    exitNodeIds,
    inputWireKind: entryWires.input,
    outputWireKind: exitWires.output,
    inputLabel: entryLabels.input,
    outputLabel: exitLabels.output,
  };
}

File: backend/app/services/pipeline_boundary.py
"""Derive pipeline boundary I/O from graph JSON."""

from __future__ import annotations

from typing import Any

from app.schemas.pipeline import PipelineBoundaryIO
from app.services.pipeline_wire_kinds import (
    BLOCKED_PIPELINE_MODELS,
    FILE_LOADER_MODELS,
    are_wire_kinds_compatible,
    get_model_wire_kinds,
    get_model_wire_labels,
)


def _parse_nodes(graph: dict[str, Any]) -> list[dict[str, Any]]:
    nodes = graph.get("nodes")
    if not isinstance(nodes, list):
        return []
    return [n for n in nodes if isinstance(n, dict) and isinstance(n.get("id"), str)]


def _parse_edges(graph: dict[str, Any]) -> list[dict[str, Any]]:
    edges = graph.get("edges")
    if not isinstance(edges, list):
        return []
    return [
        e
        for e in edges
        if isinstance(e, dict)
        and isinstance(e.get("source"), str)
        and isinstance(e.get("target"), str)
    ]


def _topological_has_cycle(
    node_ids: set[str], edges: list[dict[str, Any]]
) -> bool:
    in_degree = {nid: 0 for nid in node_ids}
    adjacency: dict[str, list[str]] = {nid: [] for nid in node_ids}

    for edge in edges:
        source = edge["source"]
        target = edge["target"]
        if source not in node_ids or target not in node_ids:
            continue
        adjacency[source].append(target)
        in_degree[target] = in_degree.get(target, 0) + 1

    queue = [nid for nid, deg in in_degree.items() if deg == 0]
    visited = 0

    while queue:
        current = queue.pop(0)
        visited += 1
        for nxt in adjacency.get(current, []):
            in_degree[nxt] -= 1
            if in_degree[nxt] == 0:
                queue.append(nxt)

    return visited != len(node_ids)


def _connected_components(
    node_ids: set[str], edges: list[dict[str, Any]]
) -> list[set[str]]:
    adjacency: dict[str, set[str]] = {nid: set() for nid in node_ids}
    for edge in edges:
        source = edge["source"]
        target = edge["target"]
        if source in node_ids and target in node_ids:
            adjacency[source].add(target)
            adjacency[target].add(source)

    visited: set[str] = set()
    components: list[set[str]] = []

    for start in node_ids:
        if start in visited:
            continue
        stack = [start]
        component: set[str] = set()
        while stack:
            current = stack.pop()
            if current in visited:
                continue
            visited.add(current)
            component.add(current)
            for neighbor in adjacency[current]:
                if neighbor not in visited:
                    stack.append(neighbor)
        components.append(component)

    return components


def _validate_edge_compatibility(
    edges: list[dict[str, Any]],
    model_by_id: dict[str, str],
    node_ids: set[str],
) -> bool:
    for edge in edges:
        source = edge["source"]
        target = edge["target"]
        if source not in node_ids or target not in node_ids:
            continue
        source_model = model_by_id.get(source, "")
        target_model = model_by_id.get(target, "")
        source_wires = get_model_wire_kinds(source_model)
        target_wires = get_model_wire_kinds(target_model)
        if not are_wire_kinds_compatible(
            source_wires["output"], target_wires["input"]
        ):
            return False
    return True


def derive_pipeline_boundary_io(graph: dict[str, Any]) -> PipelineBoundaryIO:
    nodes = _parse_nodes(graph)
    edges = _parse_edges(graph)
    errors: list[str] = []

    if not nodes:
        return PipelineBoundaryIO(valid=False, errors=["no_nodes"])

    if len(nodes) < 2:
        errors.append("insufficient_nodes")

    node_ids = {n["id"] for n in nodes}
    model_by_id = {
        n["id"]: n.get("modelId", "") for n in nodes if isinstance(n.get("modelId"), str)
    }

    for model_id in model_by_id.values():
        if model_id in BLOCKED_PIPELINE_MODELS:
            if model_id in FILE_LOADER_MODELS:
                errors.append("contains_file_loader")
            else:
                errors.append("contains_page_branch")

    if _topological_has_cycle(node_ids, edges):
        errors.append("cycle_detected")

    components = _connected_components(node_ids, edges)
    if len(components) > 1:
        errors.append("disconnected_graph")

    if not _validate_edge_compatibility(edges, model_by_id, node_ids):
        errors.append("incompatible_connection")

    in_degree = {nid: 0 for nid in node_ids}
    out_degree = {nid: 0 for nid in node_ids}
    for edge in edges:
        source = edge["source"]
        target = edge["target"]
        if source in node_ids and target in node_ids:
            out_degree[source] += 1
            in_degree[target] += 1

    entry_node_ids = [nid for nid in node_ids if in_degree[nid] == 0]
    exit_node_ids = [nid for nid in node_ids if out_degree[nid] == 0]

    if not entry_node_ids:
        errors.append("no_entry_node")
    if not exit_node_ids:
        errors.append("no_exit_node")

    entry_input_kinds: set[str] = set()
    exit_output_kinds: set[str] = set()

    for entry_id in entry_node_ids:
        entry_wires = get_model_wire_kinds(model_by_id.get(entry_id, ""))
        entry_input = entry_wires["input"]
        if entry_input in ("none", "file"):
            errors.append("invalid_entry_input")
        else:
            entry_input_kinds.add(entry_input)

    for exit_id in exit_node_ids:
        exit_wires = get_model_wire_kinds(model_by_id.get(exit_id, ""))
        exit_output = exit_wires["output"]
        if exit_output in ("none",):
            errors.append("invalid_exit_output")
        else:
            exit_output_kinds.add(exit_output)

    if len(entry_input_kinds) > 1:
        errors.append("incompatible_entry_inputs")
    if len(exit_output_kinds) > 1:
        errors.append("incompatible_exit_outputs")

    if errors:
        return PipelineBoundaryIO(
            valid=False,
            errors=errors,
            entry_node_ids=entry_node_ids,
            exit_node_ids=exit_node_ids,
        )

    primary_entry_id = entry_node_ids[0]
    primary_exit_id = exit_node_ids[0]
    entry_model = model_by_id.get(primary_entry_id, "")
    exit_model = model_by_id.get(primary_exit_id, "")

    entry_wires = get_model_wire_kinds(entry_model)
    exit_wires = get_model_wire_kinds(exit_model)
    entry_labels = get_model_wire_labels(entry_model)
    exit_labels = get_model_wire_labels(exit_model)

    return PipelineBoundaryIO(
        valid=True,
        errors=[],
        entry_node_ids=entry_node_ids,
        exit_node_ids=exit_node_ids,
        input_wire_kind=entry_wires["input"],
        output_wire_kind=exit_wires["output"],
        input_type_label=entry_labels["input"],
        output_type_label=exit_labels["output"],
    )
